File: client/move_msg.py

from message import MessageId, RecvMessage


class MoveMsg(RecvMessage):
    def __init__(self):
        super(MoveMsg, self).__init__(MessageId.MOVE)
        self.left_speed = 0
        self.left_dir = 0
        self.right_speed = 0
        self.right_dir = 0

    def from_bytes(self, data):
        self.left_speed = int.from_bytes(data[0:1], byteorder='big')
        self.left_dir = int.from_bytes(data[1:2], byteorder='big')
        self.right_speed = int.from_bytes(data[2:3], byteorder='big')
        self.right_dir = int.from_bytes(data[3:4], byteorder='big')

File: server/src/move_msg.rs
use super::message;
use message::{Message, MessageId, SendMessage};

#[derive(Debug)]
pub struct MoveMsg {
    pub id: u8,
    pub left_speed: u8,
    pub left_dir: u8,
    pub right_speed: u8,
    pub right_dir: u8,
    data: Vec<u8>,
}

impl MoveMsg {
    pub fn new() -> MoveMsg {
        let id_value = MessageId::Move as u8;
        MoveMsg {
            id: id_value,
            left_speed: 0,
            left_dir: 0,
            right_speed: 0,
            right_dir: 0,
            data: Vec::new(),
        }
    }
}

impl<'a> Message for MoveMsg {
    fn id(&self) -> u8 {
        return self.id;
    }
}

impl<'a> SendMessage for MoveMsg {
    fn size(&self) -> u32 {
        return 1 + 1 + 1 + 1;
    }

    fn to_bytes(&mut self) -> Option<&[u8]> {
        self.data.push(self.left_speed);
        self.data.push(self.left_dir);
        self.data.push(self.right_speed);
        self.data.push(self.right_dir);
        return Some(&self.data[..]);
    }
}
